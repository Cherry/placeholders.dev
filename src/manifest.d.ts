declare module '__STATIC_CONTENT_MANIFEST' {
	export const content: string;
	export default content;
}
