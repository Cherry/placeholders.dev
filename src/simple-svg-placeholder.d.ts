declare module '@cloudfour/simple-svg-placeholder' {
	type Options = {
		width?: number;
		height?: number;
		text?: string;
		dy?: string;
		fontFamily?: string;
		fontWeight?: string | number;
		fontSize?: number;
		bgColor?: string;
		textColor?: string;
		dataUri?: boolean;
		charset?: string;
	};
	export default function generateSVG(options: Options): string;
	export type { Options };
}
