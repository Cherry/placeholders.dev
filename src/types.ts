export interface Env {
	ASSETS: Fetcher;

	PLACEHOLDERS_ANALYTICS?: AnalyticsEngineDataset;
}
