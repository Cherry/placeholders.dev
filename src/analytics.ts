type AnalyticsData = {
	cached: 0 | 1;
};

export function writeDataPoint(analytics: AnalyticsEngineDataset | undefined, request: Request, data?: AnalyticsData) {
	if (!analytics) { return; }
	/* ORDER HERE IS VERY IMPORTANT. IF ANYTHING CHANGES, MUST BE APPENDED */
	const reportData = {
		blobs: [
			request.url,
			request.headers.get('user-agent'),
			request.headers.get('referer'),
			(request.cf?.httpProtocol as string) || 'invalid',
			(request.cf?.city as string) || 'unknown city',
			(request.cf?.colo as string) || 'missing colo',
			(request.cf?.country as string) || 'missing country',
			(request.cf?.tlsVersion as string) || 'invalid TLS',
		],
		doubles: [
			(request.cf?.asn as number) || 0,
			data?.cached || 0,
		],
	};
	try {
		analytics?.writeDataPoint?.(reportData);
	} catch (err) {
		console.error(err);
	}
}
