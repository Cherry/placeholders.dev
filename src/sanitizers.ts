import sanitizeHtml from 'sanitize-html';
import validateColor from 'validate-color';

export function sanitizeNumber(input: string | number) {
	const isValid = !/^\s*$/.test(String(input).trim()) && !Number.isNaN(Number(input));
	if (isValid) {
		return Number(input);
	}
	return null;
}

export function sanitizeString(input: string) {
	let value = sanitizeHtml(input, { allowedTags: [], allowedAttributes: {} });
	value = value.replaceAll(/["<>]+/g, '');
	return value;
}

export function sanitizeColor(input: string) {
	const value = sanitizeString(input); // first remove any HTML
	const isValidColor = validateColor(value);
	if (isValidColor) {
		return value;
	}
	return null;
}

export function sanitizeStringForCss(input: string) {
	let value = sanitizeString(input);
	value = value.replaceAll(/[:;]+/g, '');
	return value;
}

export function sanitizeBoolean(input: string) {
	if (String(input).toLowerCase() === 'true') {
		return true;
	}
	if (String(input) === '1') {
		return true;
	}
	return false;
}

export const sanitizers = {
	width: sanitizeNumber,
	height: sanitizeNumber,
	text: sanitizeString,
	dy: sanitizeString,
	fontFamily: sanitizeStringForCss,
	fontWeight: sanitizeNumber,
	fontSize: sanitizeNumber,
	bgColor: sanitizeColor,
	textColor: sanitizeColor,
	textWrap: sanitizeBoolean,
};
