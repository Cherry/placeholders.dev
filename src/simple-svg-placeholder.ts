// Heavily based on https://github.com/cloudfour/simple-svg-placeholder/tree/main
// License: MIT

export type Options = {
	width?: number;
	height?: number;
	text?: string;
	dy?: number;
	fontFamily?: string;
	fontWeight?: string | number;
	fontSize?: number;
	lineHeight?: number;
	bgColor?: string;
	textColor?: string;
	dataUri?: boolean;
	charset?: string;
	textWrap?: boolean;
	padding?: string;
};


export function simpleSvgPlaceholder({
	width = 300,
	height = 150,
	text = `${width}Ã—${height}`,
	fontFamily = 'sans-serif',
	fontWeight = 'bold',
	fontSize = Math.floor(Math.min(width, height) * 0.2),
	lineHeight = 1.2,
	dy = fontSize * 0.35,
	bgColor = '#ddd',
	textColor = 'rgba(0,0,0,0.5)',
	dataUri = true,
	charset = 'utf8',
	textWrap = false,
	padding = '0.5em',
}: Options = {}) {
	let content = '';
	if (textWrap) {
		content = `<foreignObject width="${width}" height="${height}">
		<div xmlns="http://www.w3.org/1999/xhtml" style="
			align-items: center;
			box-sizing: border-box;
			color: ${textColor};
			display: flex;
			font-family: ${fontFamily};
			font-size: ${fontSize}px;
			font-weight: ${fontWeight};
			height: 100%;
			line-height: ${lineHeight};
			justify-content: center;
			padding: ${padding};
			text-align: center;
			width: 100%;
		">${text}</div>
	  </foreignObject>`;
	} else {
		content = `<text fill="${textColor}" font-family="${fontFamily}" font-size="${fontSize}" dy="${dy}" font-weight="${fontWeight}" x="50%" y="50%" text-anchor="middle">${text}</text>`;
	}

	const str = `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">
	<rect fill="${bgColor}" width="${width}" height="${height}"/>
	${content}
	</svg>`;

	// Thanks to: filamentgroup/directory-encoder
	const cleaned = str
		.replaceAll(/[\t\n\r]/gim, '') // Strip newlines and tabs
		.replaceAll(/\s\s+/g, ' ') // Condense multiple spaces
		.replaceAll(/'/gim, '\\i'); // Normalize quotes

	if (dataUri) {
		const encoded = encodeURIComponent(cleaned)
			.replaceAll('(', '%28') // Encode brackets
			.replaceAll(')', '%29');

		return `data:image/svg+xml;charset=${charset},${encoded}`;
	}

	return cleaned;
}
