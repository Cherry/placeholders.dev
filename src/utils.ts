import PoPs from '@adaptivelink/pops';

/* security headers */
export const addHeaders = {
	'X-XSS-Protection': '1; mode=block',
	'X-Frame-Options': 'DENY',
	'X-Content-Type-Options': 'nosniff',
	'Referrer-Policy': 'no-referrer-when-downgrade',
	'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload',
	'Feature-Policy': [
		'geolocation \'none\';',
		'midi \'none\';',
		'sync-xhr \'none\';',
		'microphone \'none\';',
		'camera \'none\';',
		'magnetometer \'none\';',
		'gyroscope \'none\';',
		'speaker \'none\';',
		'fullscreen \'none\';',
		'payment \'none\';',
	].join(' '),
	'Content-Security-Policy': [
		'default-src \'self\';',
		'script-src \'self\' cdnjs.cloudflare.com static.cloudflareinsights.com;',
		'style-src \'self\' cdnjs.cloudflare.com \'unsafe-inline\' fonts.googleapis.com;',
		'img-src \'self\' data: images.placeholders.dev;',
		'child-src \'none\';',
		'font-src \'self\' fonts.gstatic.com cdnjs.cloudflare.com;',
		'connect-src \'self\';',
		'prefetch-src \'none\';',
		'object-src \'none\';',
		'form-action \'none\';',
		'frame-ancestors \'none\';',
		'upgrade-insecure-requests;',
	].join(' '),
} as const;

/* caching */
export const cacheTtl = 60 * 60 * 24 * 90; // 90 days
export const filesRegex = /(.*\.(ac3|avi|bmp|br|bz2|css|cue|dat|doc|docx|dts|eot|exe|flv|gif|gz|htm|html|ico|img|iso|jpeg|jpg|js|json|map|mkv|mp3|mp4|mpeg|mpg|ogg|pdf|png|ppt|pptx|qt|rar|rm|svg|swf|tar|tgz|ttf|txt|wav|webp|webm|webmanifest|woff|woff2|xls|xlsx|xml|zip))$/;

/* HTML rewriter for more accurate count of Cloudflare PoPs */
export const CloudflarePoPs = PoPs.cloudflare.code.length;
export const description = `Generate super fast placeholder images powered by Cloudflare Workers in ${CloudflarePoPs}+ edge locations.`;
export class PoPsRewriter implements HTMLRewriterElementContentHandlers {
	element(element: Element) {
		if (element.tagName === 'title') {
			element.setInnerContent(description);
		} else if (element.tagName === 'meta' && (element.getAttribute('name') === 'description' || element.getAttribute('name') === 'twitter:description' || element.getAttribute('property') === 'og:description')) {
			element.setAttribute('content', description);
		} else if (element.tagName === 'span' && element.hasAttribute('class') && element.getAttribute('class')?.includes('edgeLocations')) {
			element.setInnerContent(String(CloudflarePoPs));
		}
	}
}

export const availableImageOptions = [
	'width',
	'height',
	'text',
	'dy',
	'fontFamily',
	'fontWeight',
	'fontSize',
	'bgColor',
	'textColor',
	'darkBgColor',
	'darkTextColor',
	'textWrap',
] as const;

export function getKeys<T extends object>(obj: T) {
	return Object.keys(obj) as Array<keyof T>;
}
