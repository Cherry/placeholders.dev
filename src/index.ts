import { getAssetFromKV } from '@cloudflare/kv-asset-handler';
import manifestJSON from '__STATIC_CONTENT_MANIFEST';

import { sanitizers } from './sanitizers';
import { type Options, simpleSvgPlaceholder } from './simple-svg-placeholder';
import { type Env } from './types';
import { addHeaders } from './utils';
import {
	PoPsRewriter,
	availableImageOptions,
	cacheTtl,
	filesRegex,
} from './utils';

const DEBUG = false;
const assetManifest = JSON.parse(manifestJSON);


async function handleEvent(request: Request, env: Env, ctx: ExecutionContext) {
	const url = new URL(request.url);
	url.searchParams.sort(); // improve cache-hits by sorting search params

	const cache = caches.default; // Cloudflare edge caching
	// when publishing to prod, we serve images from an `images` subdomain
	// when in dev, we serve from `/api`
	if (url.host === 'images.placeholders.dev' || url.pathname.startsWith('/api')) {
		// do our API work
		let response;
		if (url.host === 'images.placeholders.dev') {
			response = await cache.match(url, { ignoreMethod: true }); // try to find match for this request in the edge cache
		}
		if (response) {
			// use cache found on Cloudflare edge. Set X-Worker-Cache header for helpful debug
			const newHdrs = new Headers(response.headers);
			newHdrs.set('X-Worker-Cache', 'true');
			return new Response(response.body, {
				status: response.status,
				statusText: response.statusText,
				headers: newHdrs,
			});
		}
		const imageOptions: Options = {
			dataUri: false, // always return an unencoded SVG source
			width: 300,
			height: 150,
			fontFamily: 'sans-serif',
			fontWeight: 'bold',
			bgColor: '#ddd',
			textColor: 'rgba(0,0,0,0.5)',
		};

		// options that can be overwritten
		for (const key of availableImageOptions) {
			if (url.searchParams.has(key)) {
				const rawValue = url.searchParams.get(key);
				if (!rawValue) { continue; }
				let value;
				if (key in sanitizers) {
					value = sanitizers[key](rawValue);
				} else {
					value = url.searchParams.get(key);
				}
				if (value) {
					// @ts-expect-error TODO: better type sanitizers
					imageOptions[key] = value;
				}
			}
		}
		response = new Response(simpleSvgPlaceholder(imageOptions), {
			headers: {
				'content-type': 'image/svg+xml; charset=utf-8',
				'access-control-allow-origin': '*',
			},
		});

		// set cache header on 200 response
		if (response.status === 200) {
			response.headers.set('Cache-Control', 'public, max-age=' + cacheTtl);
		} else {
			// only cache other things for 5 minutes (errors, 404s, etc.)
			response.headers.set('Cache-Control', 'public, max-age=300');
		}

		ctx.waitUntil(cache.put(url, response.clone())); // store in cache
		return response;
	}

	const getAssetPayload = {
		request: request,
		waitUntil(promise: Promise<unknown>) {
			return ctx.waitUntil(promise);
		},
	};
	const getAssetOptions = {
		ASSET_NAMESPACE: env.__STATIC_CONTENT,
		ASSET_MANIFEST: assetManifest,
	};

	// else get the assets from KV
	const options = {
		cacheControl: {
			edgeTTL: 60 * 60 * 1, // 1 hour
			browserTTL: 60 * 60 * 1, // 1 hour
			bypassCache: false,
		},
		...getAssetOptions,
	};
	if (filesRegex.test(url.pathname)) {
		options.cacheControl.edgeTTL = 60 * 60 * 24 * 30; // 30 days
		options.cacheControl.browserTTL = 60 * 60 * 24 * 30; // 30 days
	}

	let asset = null;
	try {
		if (DEBUG) {
			// customize caching
			options.cacheControl.bypassCache = true;
		}
		asset = await getAssetFromKV(getAssetPayload, options);
	} catch (err) {
		// if an error is thrown try to serve the asset at 404.html
		if (!DEBUG) {
			try {
				const notFoundResponse = await getAssetFromKV(getAssetPayload, {
					mapRequestToAsset: req => new Request(`${new URL(req.url).origin}/404.html`, req),
					...getAssetOptions,
				});

				const headers = new Headers(notFoundResponse.headers);
				headers.set('content-type', 'text/html; charset=utf-8');
				return new Response(notFoundResponse.body, {
					headers,
					status: 404,
				});
			} catch {}
		}
		const probableError = err as Error;
		return new Response(probableError?.message || probableError.toString(), { status: 500 });
	}

	if (asset && asset.headers && asset.headers.has('content-type') && asset.headers.get('content-type')?.includes('text/html')) {
		// set security headers on html pages
		for (const name of Object.keys(addHeaders)) {
			const keyedName = name as keyof typeof addHeaders;
			asset.headers.set(name, addHeaders[keyedName]);
		}
	}

	const transformed = new HTMLRewriter().on('*', new PoPsRewriter()).transform(asset);
	return transformed;
}

export default {
	async fetch(request: Request, env: Env, ctx: ExecutionContext) {
		try {
			const response = await handleEvent(request, env, ctx);
			return response;
		} catch (err) {
			console.error(err);
			return new Response('Internal Error', { status: 500 });
		}
	},
};
