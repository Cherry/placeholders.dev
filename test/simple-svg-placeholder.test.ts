import { describe, expect, test } from 'vitest';

import { type Options, simpleSvgPlaceholder } from '../src/simple-svg-placeholder';

describe('simpleSVGPlaceholder', () => {
	const testOptions: Options[] = [
		{
			dataUri: true,
			width: 100,
			height: 100,
		},
		{
			dataUri: false,
			width: 100,
			height: 100,
		},
		{
			dataUri: false,
			width: 300,
			height: 300,
			text: 'Hello World',
		},
		{
			dataUri: false,
			width: 1920,
			height: 1080,
			text: 'Hello World',
			textColor: '#fff',
			bgColor: '#000',
		},
		{
			dataUri: false,
			width: 1920,
			height: 1080,
			text: 'Hello World',
			textColor: 'rgba(255,255,255,0.5)',
			bgColor: 'rgba(0,0,0,0.5)',
		},
		{
			dataUri: false,
			width: 300,
			height: 300,
			text: 'Some super long string',
			textColor: 'rgba(255,255,255,0.5)',
			bgColor: 'rgba(0,0,0,0.5)',
		},
		{
			dataUri: false,
			width: 300,
			height: 300,
			text: 'Some super long string',
			textColor: 'rgba(255,255,255,0.5)',
			bgColor: 'rgba(0,0,0,0.5)',
			textWrap: true,
		},
	];

	test.each(testOptions)('should return accurate svg placeholder for %s', (options) => {
		const svg = simpleSvgPlaceholder(options);
		expect(svg).toMatchSnapshot();
	});
});
