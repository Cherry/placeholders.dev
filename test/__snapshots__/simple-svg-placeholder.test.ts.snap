// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`simpleSVGPlaceholder > should return accurate svg placeholder for { dataUri: false, width: 100, height: 100 } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="100" height="100" viewBox="0 0 100 100"><rect fill="#ddd" width="100" height="100"/><text fill="rgba(0,0,0,0.5)" font-family="sans-serif" font-size="20" dy="7" font-weight="bold" x="50%" y="50%" text-anchor="middle">100Ã—100</text></svg>"`;

exports[`simpleSVGPlaceholder > should return accurate svg placeholder for { dataUri: false, width: 300, height: 300, text: 'Hello World' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="300" height="300" viewBox="0 0 300 300"><rect fill="#ddd" width="300" height="300"/><text fill="rgba(0,0,0,0.5)" font-family="sans-serif" font-size="60" dy="21" font-weight="bold" x="50%" y="50%" text-anchor="middle">Hello World</text></svg>"`;

exports[`simpleSVGPlaceholder > should return accurate svg placeholder for { dataUri: false, width: 300, height: 300, text: 'Some super long string', textColor: 'rgba(255,255,255,0.5)', bgColor: 'rgba(0,0,0,0.5)' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="300" height="300" viewBox="0 0 300 300"><rect fill="rgba(0,0,0,0.5)" width="300" height="300"/><text fill="rgba(255,255,255,0.5)" font-family="sans-serif" font-size="60" dy="21" font-weight="bold" x="50%" y="50%" text-anchor="middle">Some super long string</text></svg>"`;

exports[`simpleSVGPlaceholder > should return accurate svg placeholder for { dataUri: false, width: 300, height: 300, text: 'Some super long string', textColor: 'rgba(255,255,255,0.5)', bgColor: 'rgba(0,0,0,0.5)', textWrap: true } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="300" height="300" viewBox="0 0 300 300"><rect fill="rgba(0,0,0,0.5)" width="300" height="300"/><foreignObject width="300" height="300"><div xmlns="http://www.w3.org/1999/xhtml" style="align-items: center;box-sizing: border-box;color: rgba(255,255,255,0.5);display: flex;font-family: sans-serif;font-size: 60px;font-weight: bold;height: 100%;line-height: 1.2;justify-content: center;padding: 0.5em;text-align: center;width: 100%;">Some super long string</div> </foreignObject></svg>"`;

exports[`simpleSVGPlaceholder > should return accurate svg placeholder for { dataUri: false, width: 1920, height: 1080, text: 'Hello World', textColor: '#fff', bgColor: '#000' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="1920" height="1080" viewBox="0 0 1920 1080"><rect fill="#000" width="1920" height="1080"/><text fill="#fff" font-family="sans-serif" font-size="216" dy="75.6" font-weight="bold" x="50%" y="50%" text-anchor="middle">Hello World</text></svg>"`;

exports[`simpleSVGPlaceholder > should return accurate svg placeholder for { dataUri: false, width: 1920, height: 1080, text: 'Hello World', textColor: 'rgba(255,255,255,0.5)', bgColor: 'rgba(0,0,0,0.5)' } 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="1920" height="1080" viewBox="0 0 1920 1080"><rect fill="rgba(0,0,0,0.5)" width="1920" height="1080"/><text fill="rgba(255,255,255,0.5)" font-family="sans-serif" font-size="216" dy="75.6" font-weight="bold" x="50%" y="50%" text-anchor="middle">Hello World</text></svg>"`;

exports[`simpleSVGPlaceholder > should return accurate svg placeholder for { dataUri: true, width: 100, height: 100 } 1`] = `"data:image/svg+xml;charset=utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22100%22%20height%3D%22100%22%20viewBox%3D%220%200%20100%20100%22%3E%3Crect%20fill%3D%22%23ddd%22%20width%3D%22100%22%20height%3D%22100%22%2F%3E%3Ctext%20fill%3D%22rgba%280%2C0%2C0%2C0.5%29%22%20font-family%3D%22sans-serif%22%20font-size%3D%2220%22%20dy%3D%227%22%20font-weight%3D%22bold%22%20x%3D%2250%25%22%20y%3D%2250%25%22%20text-anchor%3D%22middle%22%3E100%C3%97100%3C%2Ftext%3E%3C%2Fsvg%3E"`;
